### ISSUE TITLE
Batched requests

### ISSUE THREAD

**Issue by jennybc** [tidyverse/googledrive#126]
Many of the things we currently do with `purrr::map()` would be much faster as a batched request.

https://developers.google.com/drive/v3/web/batch

Example I just experienced: a `drive_cp()` test was creating copies exponentially (1 -> 2 -> 4 ...) so i just did a massive `drive_delete()` and it was very slow.

**Comment by MarkEdmondson1234** [comment]
I have an implementation of batching here if any use:
https://github.com/MarkEdmondson1234/googleAuthR/blob/master/R/googleAuthR_batch.R#L265

You have to make a POST request with all your requests as content within it, it was very picky about whitespace too.

**Comment by jennybc** [comment]
Thanks @MarkEdmondson1234! It's good to know you've got it working and to see a working example. This would go over in gargle, I believe, because AFAICT the batched request framework is common across multiple Google APIs.

Did you see the speed gains you were expecting / hoping for?

**Comment by MarkEdmondson1234** [comment]
Sure, take or leave the code but at least a working example to compare against. 

It's really good for the APIs that support it, although not all do - google analytics v3 does but v4 doesn't for example. 

There are also different API limits for each application for concurrent requests, but you can batch up to them, so for example 10 requests per time for google analytics v3, but 1000 for search console - big gains there. 

You can also combine requests from different API in one batch request, such as googlesheets and google drive, (BigQuery?), although no guarantee they execute in the same order you send them in.

**Comment by DataStrategist** [comment]
Sorry for Issue-rezzing, but I think this would help a problem I'm having recently. As of a few days ago I'm getting `Error 500`  problems when I try to `drive_share` a file to multiple people. It's a strange and intermittent bug that shouldn't be happening, I'm sending on the order of 30 share-invites per day... wouldn't trigger any limits on google's API. Anyway, I think if I could bulk-share I think this problem would go away.

FWIW, here's my code, not pretty but tends to work:

```
permissionerGuts <- function(xx, yy) {
  a <- drive_share(
    file = xx, role = "writer", type = "user",
    emailAddress = yy,
    emailMessage = "Please let me know !"
  )
  paste("OK for: ", yy, " getting file ", a$name)
}

## and call the function in a map
boom <- googledrive::drive_get(FileNameWithoutExtension)
Peeps <- c("email1@test.com", "email2@test.com")
purrr::map2_chr(.x = rep(list(googledrive::as_id(boom$id)), length(Peeps)), 
                .y = as.list(Peeps), .f = permissionerGuts)
```

I'm on version 1.3 of googledrive

**Comment by jennybc** [comment]
@DataStrategist Can you show an actual error? I'm just curious if it's exactly 500 or something "in the 500s" and what the other details are.

**Comment by jennybc** [comment]
The dev version of googledrive uses a better (more verbose) error handler from gargle, so if you're in a position to try that, it could be revealing. But there are some small-but-not-zero auth changes, I should warn you.

**Comment by DataStrategist** [comment]
Sorry for the delay, I just got married and therefore haven't been keeping up w/ my github issues :) Still, apologies for being crappy.

Update! It turns out that those error 500s were due to google being suspicious about my activity... they ultimately decided that the emails I was sending were spam and permanently disabled that account! I have no idea why, I was sending non-commercial emails to a list of employees and partners that specifically opted in... but they made their final decision.

So yeah, it had nothing to do with you. Answering your specific question, it was EXACTLY an error 500. I can't find the specific error in my logs because it was breaking my script so it wasn't captured by any of my `safely()` traps.
